### ISSUE TITLE
Trying to get the driver status

### ISSUE THREAD

**Issue by Clabberlang** [kens-git/f122_telemetry_parser#12]
hi! before i ask the question, just want to let you know im kinda new to coding and programming in general.

so i tried to get the status of the driver e,g in lap etc, but seems to not be able to figure it out. i tried to print it directly with 
``` print(LapDataData.currentLapTimeInMS)
the output is still not being parsed. can you guide me or let me know things that i have to do. much appreciate it

**Comment by kens-git** [comment]
@Clabberlang I think the formatting of your comment got messed up because I can't see any code.

Also, are you running F1 22 or F1 23? I haven't modified this at all to work with F1 23 so that could be an issue if you're running this year's game.

**Comment by Clabberlang** [comment]
its f1 22,   print(LapDataData.currentLapTimeInMS)

that was the code that i tried to see the value off it, but it seems it still in a raw state of data with returning as c-types , ofs etc.. so i kinda dont know where to go from that

**Comment by kens-git** [comment]
@Clabberlang Here's an example of filtering for the current lap time:

```
def filter_lap_data(self, packet: LapDataPacket):
        data = packet.lapData
        driver_lap_data = data[0]
        print(driver_lap_data.currentLapTimeInMS)
```

With your code you appear to be printing out the _definition_ of the field itself, which probably gives you: `<Field type=c_ulong, ofs=4, size=4>`. The difference between my code and yours is that I'm printing out an _instance_ of currentLapTimeInMS, not the _definition._ The short explanation is basically that the definition is just a description of what values the variable can or should hold (in this case, an unsigned integer of 4 bytes), and an instance is a 'real' version with actual values, created using the definition as a template.

In this example `packet.lapData` is a list of LapDataData instances that I store in `data`, where each item in that list is lap data for a specific driver. `driver_lap_data` then takes the first driver's data from that list and stores it. The next line then prints out the `currentLapTimeInMS` for that driver's current lap.

**Comment by Clabberlang** [comment]
thanks for the reply. just one more question, is there any documentation that i can read or follow to be the guide for all of this stuff? much appreciate it

**Comment by kens-git** [comment]
You'lll need to know the basics of Python pretty well, up to classes and inheritance since that's how the filters are implemented.

If you want to understand what's in the packets you'll need to know what each data type is. Since the game is most likely written in C++ (a superset of C), this can be used as a guide for figuring out what those types are: https://os.mbed.com/handbook/C-Data-Types

After that, the packet specification tells you what's in each packet and provides some other helpful information:
https://answers.ea.com/t5/General-Discussion/F1-22-UDP-Specification/m-p/11551319?attachment-id=657933

As an example, looking at the PacketParticipantsData struct in the spec we can see it's made up of:
```
PacketHeader m_header; // Header
uint8 m_numActiveCars; // Number of active cars in the data – should match number of
// cars on HUD
ParticipantData m_participants[22];
```
and comparing that to the packet I've defined in this project:
```
class ParticipantsPacket(Packet):
    _fields_ = [
        ('numActiveCars', c_uint8),
        ('participants', ParticipantsData * GRID_SIZE),
    ]
```
The PacketHeader is included in every packet and in this project every packet inherits from a 'Packet' class which represents this header, which then makes it available to every packet in our program. In general, since each packet type is delivered to a specific filtering function in our filters we won't care about what's in the packet header and we can ignore it. We can also ignore the 'm_' and 'c_' prefixes in the names for now: they serve as a way to distinguish the names of these things from other names but the specifics of that aren't important right now.

Next, we have numActiveCars. We can see in the spec that it's defined as a 'uint8', meaning an 8 bit integer that is unsigned. Looking at the C types guide I linked we can see that this represents a value that can be anything from 0 up to 255, and we can then always expect when we access this value it will be in that range.

Next, we have participants. In the spec it's defined as an array (roughly equivalent to a tuple in Python) of 'ParticipantData' instances with a fixed size/length of 22. The syntax isn't obvious but in this project `ParticipantsData * GRID_DATA` is defining a ctypes array (think: tuple) of ParticipantsData instances that has a size of `GRID_SIZE`. Looking into the project's code you'll see that `GRID_SIZE` is a constant that is given the value 22. So, whenever we access the 'participants' field of a ParticipantsPacket instance, we know we're getting an array of 22 ParticipantsData instances. Note that not every instance will contain valid values, since something like a time trial will only have 1 driver. In this case the rest of the array will still be there but should be ignored.

From here, the rest of the fields in the packets should make sense. The only other trick may be the arrays of 'chars' that are used for text strings. In the spec you might see something like this: `char m_name[48]`, which is an array of text characters that has a length 48. In Python we usually only ever think of 'arrays of characters' as strings, but C/C++ is different in this regard so there's a little more work involved when dealing with these arrays of characters. Fortunately, if you look in the utilities folder you'll see a 'data.py' file which contains a 'to_string' function. You can import this function and pass any character arrays from the packets into it and it'll return the Python string equivalent which is much easier to work with.

In short, without knowing the basics of Python well it will be very difficult to make sense of this stuff, so definitely go through a book or tutorial at least until you've reached the section on classes and inheritance and understand those concepts. Once you've covered that, making sense of the rest of this should be simpler.

**Comment by Clabberlang** [comment]
sorry if to bother, but i got another question, how do you access other filter function on a different filter function, what i mean is that if im in fiter_participants, how do i access data from filter_session so i can use it under the filter_participants section. also can you explain `articipant = cast(Array[ParticipantsData],
                           self.participants)[vehicle_index]` this part of  your code? i look it up online but cant seem to figure it out.

TIA, also if you mind answering these question, you can close the ticket, idm

**Comment by kens-git** [comment]
Check out the LogFilter constructor (the ```__init__``` method) and the LogFilter filter_participants method to see how that is handled. Basically, when you first receive the participants data in the participants filter function you would store it in an instance variable in that class instance that you can access in the other filter functions. Note that the way I coded it in the LogFilter may not work with online races because my implementation only takes the first participant packet and stores the data in it and ignores any other participant packet after that. For online, it's possible that if people leave the race/lobby then the participant data will change over the course of the race and you may want to respond to those changes in some way.

The line with the ```cast``` is an attempt to force a more strongly typed style into the code. For a language like C++, anywhere a variable is referred to the type of it must be known (e.g., declaring an integer looks like: ```int num = 10;```, and anywhere ```num``` is used it is known that it's an integer), where in a loosely typed language like Python these types are optional (and a declaration is just: ```num = 10```). For that cast, I'm explicitly saying that self.participants is being type-cast as an array of ParticipantsData instances. I can't remember the details of some of this because it's been a while since I looked at this project, but this may help the IDEs autocomplete in certain cases, and is generally meant to clarify exactly what type a variable is expected to be.
